### PR TITLE
doc: Input ctors behaviour

### PR DESCRIPTION
**Pull request by benmccallum**

### CODE REVIEW

File: website/src/docs/hotchocolate/defining-a-schema/input-object-types.md [hunk 1]
```diff
@@ -41,6 +41,23 @@ public class Mutation
 
 > Note: If a class is used as an argument to a resolver and it does not end in `Input`, Hot Chocolate (by default) will append `Input` to the type name in the resulting schema.
 
+If you would like your input type classes to be immutable, or you are using [nullable reference types](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references), you can add a constructor and Hot Chocolate will use it under the hood. Just note that, (1) the type of the argument must exactly match the property's type, and, (2) the name of the argument must match the property name (bar a lowercase first letter).
```
Comment (benmccallum) [inline website/src/docs/hotchocolate/defining-a-schema/input-object-types.md:44]: Would like to add a 3, but not sure on the behaviour. Let me know and we can GTM.

(3) you should provide arguments for every property that doesn't have a setter?
(e.g. available setters are called when a ctor is used)

-or-
(3) you **must** provide arguments for every property
(e.g. available setters are NOT called when a ctor is used?)

File: website/src/docs/hotchocolate/defining-a-schema/input-object-types.md [hunk 2]
```diff
@@ -41,6 +41,23 @@ public class Mutation
...
+    public string Title { get; }
+    public string Author { get; }
+  
+    public BookingInput(string title, string author)
+    {
+        Title = title;
+        Author = author;
+    }
+}
+```
```
Comment (benmccallum) [inline website/src/docs/hotchocolate/defining-a-schema/input-object-types.md:59]:
```suggestion
```

You can even use records! The equivalent of above would be:

```csharp
public record BookingInput(string Title, string Author);
```
```

File: website/src/docs/hotchocolate/defining-a-schema/input-object-types.md (hunk 1 above)
Comment (michaelstaib) [reply website/src/docs/hotchocolate/defining-a-schema/input-object-types.md:44]: (3) you must provide arguments for every property
(e.g. available setters are NOT called when a ctor is used?)

Comment (benmccallum) [reply website/src/docs/hotchocolate/defining-a-schema/input-object-types.md:44]: Literally just hit the exception right now and got this haha:
```
System.InvalidOperationException: 
No compatible constructor found for input type type \u0060BookingService.Features.BookingOffers.RespondToBookingOfferInput\u0060.\r\n
Either you have to provide a public constructor with settable properties 
or a public constructor that allows to pass in values for read-only properties.
There was no way to set the following properties: .\r\n
```

Bit strange how it looks correct and the list is empty though! Will figure it out

Comment (benmccallum) [reply website/src/docs/hotchocolate/defining-a-schema/input-object-types.md:44]:
```suggestion
If you would like your input type classes to be immutable, or you are using [nullable reference types](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references), you can provide a non-empty constructor and Hot Chocolate will instead use that when instantiating the input. Just note that, (1) the type of the argument must exactly match the property's type, (2) the name of the argument must match the property name (bar a lowercase first letter), and (3) no setters will be called, so you need to provide arguments for all the properties. 

Hot Chocolate even validates any custom input constructor at schema build time, so you don't need to worry about breaking things during refactoring!
```

**Review by michaelstaib — DISMISSED** [review]

**Pull request merged**